hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/NativeSparkDataSet.java

```diff
@@ -623,7 +623,12 @@ private Dataset getDataset(DataSet dataSet) throws StandardException {
             Dataset<Row> right = getDataset(dataSet);
 
             //Do the subtract
-            Dataset<Row> result = left.except(right);
+            Dataset<Row> result;
+            if(isDistinct) {
+                result = left.distinct().except(right.distinct());
+            } else {
+                result = left.except(right);
+            }
```

Comment:
I think it'd be better to generate a distinct node during query compilation, that way we'd be able to push it down to the table scan. Nowadays I think we don't do anything fancy with it, it's implemented as a Scan + Distinct, but in the future we could use a Skipping HBase filter similarly to what we do with IndexPrefixIteratorMode

In short, I think this is a good change, but in the future it could be improved a little bit.

Reply:
Thanks Daniel! I thought I would attempt to keep it localized to Spark for two reasons:

- As discussed before, it looks to me like a bug in Spark `except` operation itself and it should be fixed there. With this PR I tried to isolate the fix as much as possible.
- If we have no index / PK, then there is not much difference between `DistinctScanOperation` + `DataSet.subtract` and `ScanOperation` + `DataSet.subtract`, both iterate on the entire table but use an intermediary HashSet when iterating over the values to get "notified" about a distinct value. I am guessing if there is a more efficient keyed AccessPath then optimizer should choose that anyway which can be iterated upon with a normal iterator or a skip iterator,, but I haven't tested that to be honest.